### PR TITLE
WIP: [daint] Julia easyblock: insert hostname dynamically at module load time

### PR DESCRIPTION
**Pull request by omlins**
This PR makes the insertion of the hostname into the environment variables set by the Julia module dynamic: while previously it was done at build time it is now done at module load time.

### CODE REVIEW

File: easybuild/easyblocks/julia.py [hunk 1]
```diff
@@ -56,8 +56,9 @@ def extra_options(extra_vars=None):
     def get_environment_folder(self):
         env_path = ''
 
-        hostname = socket.gethostname()
-        hostname_short = ''.join(c for c in hostname if not c.isdigit())
+        #hostname = socket.gethostname()
+        #hostname_short = ''.join(c for c in hostname if not c.isdigit())
```
Comment (gppezzi) [inline easybuild/easyblocks/julia.py:60]: I suggest to omit the previous implementation from the code, we can refer to git if we need old versions.

**Comment by omlins** [comment]
Environment variables set when called from Dom:
```
omlins@dom101:/apps/daint/UES/omlins/julia-1.5.0/production> module show /apps/daint/UES/omlins/julia-1.5.0/modules/all/Julia/1.5.0-CrayGNU-20.08 2>&1 | grep -i mc
setenv		 JULIA_MPIEXEC_ARGS -C mc 
setenv		 JULIA_PROJECT ~/.julia/1.5.0/dom-mc/environments/1.5.0-dom-mc 
setenv		 JULIA_DEPOT_PATH ~/.julia/1.5.0/dom-mc:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/extensions:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/local/share/julia:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/share/julia 
setenv		 EBJULIA_USER_DEPOT_PATH ~/.julia/1.5.0/dom-mc 
setenv		 JULIA_LOAD_PATH @:@#.#.#-dom-mc:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/extensions/environments/1.5.0-dom-mc:@stdlib 
setenv		 EBJULIA_USER_LOAD_PATH @:@#.#.#-dom-mc 
setenv		 EBJULIA_ADMIN_LOAD_PATH /apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/extensions/environments/1.5.0-dom-mc:@stdlib 
setenv		 EBJULIA_ENV_NAME 1.5.0-dom-mc 
```

Environment variables set when called from Daint:
```
omlins@daint102:~/check_julia_modules> module show /apps/daint/UES/omlins/julia-1.5.0/modules/all/Julia/1.5.0-CrayGNU-20.08 2>&1 | grep -i mc
setenv		 JULIA_MPIEXEC_ARGS -C mc 
setenv		 JULIA_PROJECT ~/.julia/1.5.0/daint-mc/environments/1.5.0-daint-mc 
setenv		 JULIA_DEPOT_PATH ~/.julia/1.5.0/daint-mc:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/extensions:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/local/share/julia:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/share/julia 
setenv		 EBJULIA_USER_DEPOT_PATH ~/.julia/1.5.0/daint-mc 
setenv		 JULIA_LOAD_PATH @:@#.#.#-daint-mc:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/extensions/environments/1.5.0-daint-mc:@stdlib 
setenv		 EBJULIA_USER_LOAD_PATH @:@#.#.#-daint-mc 
setenv		 EBJULIA_ADMIN_LOAD_PATH /apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08/extensions/environments/1.5.0-daint-mc:@stdlib 
setenv		 EBJULIA_ENV_NAME 1.5.0-daint-mc 
```

**Review by gppezzi — CHANGES_REQUESTED** [review]

**Comment by omlins** [comment]
This approach works perfectly well for Julia alone. However, as is it does not work for the additionally installed packages as their installation also relies on the hostname. It could theoretically be changed, but it is quite complex and needs changes in all the easyblocks. I am not sure it is worth it just to prevent this glitch.

**Comment by omlins** [comment]
I think a good solution could be to keep the admin package installation using the static hostname and make only the user installations use a dynamic hostname (by using the dynamic hostname only for the EBJULIA_USER_DEPOT_PATH and the EBJULIA_USER_LOAD_PATH).

**Comment by omlins** [comment]
Digging into it I considered in the end easier to make the admin installation hostname independent. So, the user package installation related paths use a dynamic hostname and the admin package installation related paths drop the hostname (see `EBJULIA_ADMIN_LOAD_PATH`).

Module when loaded from Daint:
```
omlins@daint104:~> module show /apps/daint/UES/omlins/julia-1.5.0/modules/all/Julia/1.5.0-CrayGNU-20.08-cuda 2>&1 | grep gpu
setenv		 JULIA_MPIEXEC_ARGS -C gpu 
setenv		 JULIA_PROJECT ~/.julia/1.5.0/daint-gpu/environments/1.5.0-daint-gpu 
setenv		 JULIA_DEPOT_PATH ~/.julia/1.5.0/daint-gpu:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/extensions:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/local/share/julia:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/share/julia 
setenv		 EBJULIA_USER_DEPOT_PATH ~/.julia/1.5.0/daint-gpu 
setenv		 JULIA_LOAD_PATH @:@#.#.#-daint-gpu:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/extensions/environments/1.5.0-gpu:@stdlib 
setenv		 EBJULIA_USER_LOAD_PATH @:@#.#.#-daint-gpu 
setenv		 EBJULIA_ADMIN_LOAD_PATH /apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/extensions/environments/1.5.0-gpu:@stdlib 
setenv		 EBJULIA_ENV_NAME 1.5.0-gpu 
```

Module when loaded from Dom:
```
omlins@dom101:/apps/daint/UES/omlins/julia-1.5.0/production> module show /apps/daint/UES/omlins/julia-1.5.0/modules/all/Julia/1.5.0-CrayGNU-20.08-cuda 2>&1 | grep gpu
setenv		 JULIA_MPIEXEC_ARGS -C gpu 
setenv		 JULIA_PROJECT ~/.julia/1.5.0/dom-gpu/environments/1.5.0-dom-gpu 
setenv		 JULIA_DEPOT_PATH ~/.julia/1.5.0/dom-gpu:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/extensions:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/local/share/julia:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/share/julia 
setenv		 EBJULIA_USER_DEPOT_PATH ~/.julia/1.5.0/dom-gpu 
setenv		 JULIA_LOAD_PATH @:@#.#.#-dom-gpu:/apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/extensions/environments/1.5.0-gpu:@stdlib 
setenv		 EBJULIA_USER_LOAD_PATH @:@#.#.#-dom-gpu 
setenv		 EBJULIA_ADMIN_LOAD_PATH /apps/daint/UES/omlins/julia-1.5.0/software/Julia/1.5.0-CrayGNU-20.08-cuda/extensions/environments/1.5.0-gpu:@stdlib 
setenv		 EBJULIA_ENV_NAME 1.5.0-gpu
```

**Comment by omlins** [comment]
The stacked environment works.

**Review by twrobinson — COMMENTED** [review]
Hi @omlins Is this ready for review? If so please add the tag. If it is ready, is there a way for me to test the modules (e.g. modules in scratch or an apps sandbox). cheers

**Comment by omlins** [comment]
retest this please

**Comment by omlins** [comment]
@twrobinson and @gppezzi : I did a lot of testing and from my point of view it is ready to be merged now.

**Comment by omlins** [comment]
@twrobinson 
To test the modules I have built do:
```
module load daint-gpu
module use /apps/daint/UES/omlins/julia-1.5.0/modules/all/
```

**Comment by twrobinson** [comment]
> The stacked environment works (I have tested it) and both the `Julia` and the `JuliaExtensions` module produce the right variables.

I don't think it works on a compute node. I get the following - can you check again? 

(@1.5.0-nid-gpu) pkg> add Example
  Resolving package versions...
  Installed Example ─ v0.5.3
Updating `~/.julia/1.5.0/nid-gpu/environments/1.5.0-nid-gpu/Project.toml`
  [7876af07] + Example v0.5.3
Updating `~/.julia/1.5.0/nid-gpu/environments/1.5.0-nid-gpu/Manifest.toml`
  [7876af07] + Example v0.5.3

(@1.5.0-nid-gpu) pkg>

**Comment by omlins** [comment]
> I don't think it works on a compute node. I get the following - can you check again?

Yes, the installation of package works only from the login nodes as the Pkg manager fails to access git from the compute nodes. This is the same with the current version and I don't think there is a straightforward solution to this.

**Comment by twrobinson** [comment]
> > I don't think it works on a compute node. I get the following - can you check again?
> 
> Yes, the installation of package works only from the login nodes as the Pkg manager fails to access git from the compute nodes. This is the same with the current version and I don't think there is a straightforward solution to this.

No, the package manger works fine from the compute node. The issue is the package is installed in "nid".

**Comment by omlins** [comment]
Oh, I see. This means that it is still using a dynamic hostname at a place it should not.

**Comment by victorusu** [comment]
@omlins is this PR still makes sense? I think we have changed the way we have the JuliaBlocks, now, right? 
What is the impact on JupyterLab's Julia integration?

**Comment by omlins** [comment]
Yes, this can be closed as this approach of dynamically inserting the hostname is problematic and I found a better approach.

**Pull request closed without merging**